### PR TITLE
HC-Tree 1

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
### Details
This branch does not fully fix HC-Tree, but it's been used to diagnose two HC-Tree bugs in sqlite, and it fixes one existing bug in Bedrock that is much more frequent when using HC-Tree.

The fix for the existing bug is honestly more of a workaround than a real fix, but it seems to help at least in dev, and we do see this issue in production occasionally, so hopefully this resolves that issue there as well.

I'll explain:

When we start the `_replicate` function in a new thread, there is no guarantee that the thread starts at any particular time. In general, these threads start "immediately" which means nothing concrete but is within a few milliseconds of when we tried to start the thread. However, when we try to start many of these threads at once (where "many" seems to mean several thousand on production, or several hundred on dev), which is to say when we're falling behind on replication, these start. not only out-of-order or delayed, but it seems that occasionally, one thread fails to start at all, or gets pushed perpetually to the bottom of a queue somewhere.

It's feasible that this thread might eventually actually run and finish, but because replicate threads need to finish in a specific order, if this thread does not finish, then none of the threads started after it will ever finish either, and so the number of threads continues to grow, exacerbating the problem.

This change forces the sync thread to wait after each new `_replicate` thread is instantiated, just until it actually starts running (but not until it's done any real work) before starting the next `_replicate` thread. This means that each thread has definitely been scheduled to run and actually started before the next one will be started. It's feasible this may be slightly slower, but it also seems to resolve this problem.

A second and more minor change here is to not use `ResourceMonitorThread` here, and instead just use `thread`. I feel like `ResourceMonitorThread` is potentially dangerous to use with `detach`ed threads (even though I do think the current implementation works OK) because the actual `ResourceMonitorThread` is (in this case, and probably the most typical case) destroyed well before the thread completes. Using any non-thread-local variables in `ResourceMonitorThread` would generally cause a segfault. I don't think we strictly need this here, so I'm removing it for clarity. We could always add it back if required.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/337537

### Tests
No new tests, though the HC-Tree tests do pass when used with this code.
_________
**Internal Testing Reminder:** when changing bedrock, please compile auth against your new changes

### CODE REVIEW

**Review by flodnv — APPROVED** [review]

File: sqlitecluster/SQLiteNode.cpp [hunk 1]
```diff
@@ -1645,7 +1652,17 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
...
                     uint64_t threadAttemptStartTimestamp = STimeNow();
-                    ResourceMonitorThread([=, this](){this->_replicate(peer, message, _dbPool->getIndex(false), threadAttemptStartTimestamp);}).detach();
+                    _replicateThreadStarted = false;
+                    thread(&SQLiteNode::_replicate, this, peer, message, _dbPool->getIndex(false), threadAttemptStartTimestamp).detach();
+                    {
+                        unique_lock<mutex> lock(_replicateStartMutex);
+                        while (!_replicateThreadStarted) {
+                            _replicateStartCV.wait(lock);
+                            if (!_replicateThreadStarted) {
+                                SINFO("condition variable finished waiting but replicate thread not started.");
```
Comment (danieldoglas) [inline sqlitecluster/SQLiteNode.cpp:1662]: Just for my own education: is there any situation where this might be logged? Considering this will always wait on the notification on _replicateStartCV, and we're not using a timeout here, the only situation where we would leave this while loop is if the thread starts and _replicateThreadStarted is set to true, right?

Comment (danieldoglas) [reply sqlitecluster/SQLiteNode.cpp:1662]: Also, could it happen that `_replicateStartCV.notify_all();` runs before we reach the line `_replicateStartCV.wait(lock);` above?

**Review by danieldoglas — APPROVED** [review]
LGTM, just a few questions

File: sqlitecluster/SQLiteNode.cpp [hunk 2]
```diff
@@ -1645,7 +1652,17 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
                 SDEBUG("Spawning concurrent replicate thread (blocks until DB handle available): " << threadID);
                 try {
                     uint64_t threadAttemptStartTimestamp = STimeNow();
-                    ResourceMonitorThread([=, this](){this->_replicate(peer, message, _dbPool->getIndex(false), threadAttemptStartTimestamp);}).detach();
+                    _replicateThreadStarted = false;
```
Comment (danieldoglas) [inline sqlitecluster/SQLiteNode.cpp:1655]: do we need this to be atomic?

Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:1655]: Given that only one replicate thread can try and change this at a time, and we haven't started the next one to do it yet, nothing else will be changing this, so we don't need to.

File: sqlitecluster/SQLiteNode.h [hunk 3]
```diff
@@ -6,6 +6,9 @@
 #include <sqlitecluster/SQLitePool.h>
 #include <sqlitecluster/SQLiteSequentialNotifier.h>
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.h:8]:
```suggestion
```

File: sqlitecluster/SQLiteNode.h [hunk 4]
```diff
@@ -349,6 +352,11 @@ class SQLiteNode : public STCPManager {
     // Remove. See: https://github.com/Expensify/Expensify/issues/208449
     atomic<int> _priority;
 
+    // These three variables are used to coordinate the startup of replication threads to guarantee each thread starts before we attempt to start the next one.
+    mutex _replicateStartMutex;
+    condition_variable _replicateStartCV;
```
Comment (cead22) [inline sqlitecluster/SQLiteNode.h:357]:
```suggestion
    condition_variable _replicateStart;
```

File: sqlitecluster/SQLiteNode.cpp (hunk 1 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:1662]: It's possible because of this: https://en.wikipedia.org/wiki/Spurious_wakeup

See the `condition_variable` documentation here:
https://en.cppreference.com/w/cpp/thread/condition_variable

Specifically:
> Call [wait](https://en.cppreference.com/w/cpp/thread/condition_variable/wait), [wait_for](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for), or [wait_until](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_until) on the std::condition_variable (atomically releases the mutex and suspends thread execution until the condition variable is notified, a timeout expires, or a [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) occurs, then atomically acquires the mutex before returning).

> Also, could it happen that _replicateStartCV.notify_all(); runs before we reach the line _replicateStartCV.wait(lock); above?

Yes, but in that case we will never get to that line, `_replicateThreadStarted` will already be true when we get to the `while` loop.

See the notes here for some detail:
https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

You get to read the only use I've ever come across of the word "pessimization".

**Review by cead22 — APPROVED** [review]

File: sqlitecluster/SQLiteNode.h (hunk 4 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.h:357]: I don't get why we'd remove the type info from the name of the condition variable but not the mutex. They are two coordinated variables for the same purpose.

**Review by mjasikowski — APPROVED** [review]

File: sqlitecluster/SQLiteNode.cpp (hunk 1 above)
Comment (danieldoglas) [reply sqlitecluster/SQLiteNode.cpp:1662]: I see, thanks.

 The reason why you didn't use a predicate by overloading the wait method here is just so you could have that SINFO, or is it preferable to write that while loop on the condition?

File: sqlitecluster/SQLiteNode.h (hunk 4 above)
Comment (cead22) [reply sqlitecluster/SQLiteNode.h:357]: I would've suggested removing Mutex from the other var name too, I just missed it

Comment (flodnv) [reply sqlitecluster/SQLiteNode.h:357]: I think this is clearer
